server/src/main/java/org/elasticsearch/index/seqno/GlobalCheckpointSyncAction.java

```diff
@@ -77,24 +78,26 @@ protected void shardOperationOnPrimary(
         IndexShard indexShard,
         ActionListener<PrimaryResult<Request, ReplicationResponse>> listener
     ) {
-        ActionListener.completeWith(listener, () -> {
-            maybeSyncTranslog(indexShard);
-            return new PrimaryResult<>(request, new ReplicationResponse());
-        });
+        maybeSyncTranslog(indexShard, listener, () -> new PrimaryResult<>(request, new ReplicationResponse()));
     }
 
     @Override
     protected void shardOperationOnReplica(Request shardRequest, IndexShard replica, ActionListener<ReplicaResult> listener) {
-        ActionListener.completeWith(listener, () -> {
-            maybeSyncTranslog(replica);
-            return new ReplicaResult();
-        });
+        maybeSyncTranslog(replica, listener, ReplicaResult::new);
     }
 
-    private static void maybeSyncTranslog(final IndexShard indexShard) throws IOException {
+    private static <T> void maybeSyncTranslog(IndexShard indexShard, ActionListener<T> listener, CheckedSupplier<T, Exception> supplier) {
         if (indexShard.getTranslogDurability() == Translog.Durability.REQUEST
             && indexShard.getLastSyncedGlobalCheckpoint() < indexShard.getLastKnownGlobalCheckpoint()) {
-            indexShard.sync();
+            indexShard.syncGlobalCheckpoint(indexShard.getLastKnownGlobalCheckpoint(), e -> {
```

Comment:
I think we can be a little more conservative by capturing the GCP to sync in `org.elasticsearch.index.shard.IndexShard#maybeSyncGlobalCheckpoint` rather than using the last-known one captured here. But that's a bigger (and kinda unrelated) change, so it can be a follow-up.

Reply:
Would prefer follow-up here.